Add release announcement for v1.32 (#48666)
kubernetes/website: content/en/blog/_posts/2024-12-11-Kubernetes-v1-32-Release.md

Reviewer: Overall style nit: we write headings in sentence case

eg “Recover from volume expansion failure” (or “Recovery from volume expansion failure“), not “Recover from Volume Expansion Failure“

Author: Fixed!

Reviewer 2: Wrap lines at 80 characters (although that might not apply to blogs as strictly)

Reviewer: Wrap point can be 80, can be 120. Shouldn't be more than 200.

Author: Wrapped in the latest commit at 120, I will wrap again before getting the final labels in case it's needed.

I personally do not agree wrapping is anyhow relevant - though written [in the style guide](https://kubernetes.io/docs/contribute/style/style-guide/#line-breaks) - as I don't see any tangible benefit since this is not code but text.

That said, I'm fine doing it since it's an operation faily easy to automate - with some errors possible (see the last commits)

Reviewer: > I personally do not agree wrapping is anyhow relevant - though written [in the style guide](https://kubernetes.io/docs/contribute/style/style-guide/#line-breaks) - as I don't see any tangible benefit since this is not code but text.

Actual people still work with the Markdown source.
If you do it, reviewers and several of our 13 localization teams will thank you for making their work less difficult.

Reviewer: One more thing. This needs to be `content/en/blog/_posts/2024-12-11-kubernetes-v1-32-release/index.md` (in a directory) so that there is a place to put the release logo.

Author: Yep, I thought about that, let me fix it before adding the logo so the other PR will be cleaner.

Similar to previous releases, the release of Kubernetes v1.32 introduces new stable, beta, and alpha features. 
The consistent delivery of high-quality releases underscores the strength of our development cycle and the vibrant support from our community.
This release consists of 44 enhancements.
Of those enhancements, 13 have graduated to Stable, 12 are entering Beta, and 19 have entered in Alpha.

## Release theme and logo (WIP)
{{< figure src="/images/blog/2024-12-11-kubernetes-1.32-release/k8s-1.32.png" alt="Kubernetes v1.32 logo" class="release-logo" >}}

<TODO upload image to static/images/blog/2024-12-11-kubernetes-1.32-release/k8s-1.32.png>

The Kubernetes v1.32 Release Theme is "{release-name}".

Kubernetes v1.32's {release-story} 

### A note on DRA enhancements

In this release, like the previous one, the Kubernetes project continues proposing a number of enhancements to the Dynamic Resource Allocation (DRA), a key component of the Kubernetes resource management system. These enhancements aim to improve the flexibility and efficiency of resource allocation for workloads that require specialized hardware, such as GPUs, FPGAs and network adapters.

> The core part of Dynamic Resource Allocation (DRA), the [structured parameter support](https://github.com/kubernetes/enhancements/issues/4381), got promoted to beta.

Reviewer 3: ```suggestion
The core part of Dynamic Resource Allocation (DRA), the [structured parameter support](https://github.com/kubernetes/enhancements/issues/4381), got promoted to beta.
Publishing information about a device in the ResourceClaim status, in particular for network interfaces, got added as part of KEP [#4817](https://github.com/kubernetes/enhancements/issues/4817) as a new alpha feature.
```

Resource health status has been added in the Pod status, as outlined in KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680). This feature provides a more detailed view of the health status of resources in the Pod, allowing users to better understand the state of their resources and take appropriate action when necessary.

> KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816) adds support for prioritized selection criteria associated to a device request in a ResourceClaim.

Reviewer 3: ```suggestion
Another new KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816) will add support for prioritized selection criteria associated to a device request in a ResourceClaim.
```

Same here, implementation is pending.

This work was done as a part of [KEP #3962](https://github.com/kubernetes/enhancements/issues/3962) by [SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery).

### Pod-Level Resource Specifications 
    
This enhancement introduces a simplified way to manage resources in Kubernetes by allowing users to define resource requests and limits for an entire pod instead of specifying them individually for each container. This creates a shared resource pool that all containers in the pod can dynamically utilize, optimizing resource usage and reducing over-provisioning, especially for workloads with varying or bursty resource demands.

With this feature, users gain more flexibility and efficiency in resource allocation, as the pod-level settings complement existing container-level specifications. By directly leveraging pod-level cgroup settings, Kubernetes ensures better performance for tightly coupled applications while maintaining backward compatibility with container-level configurations. This enhancement represents a significant step forward in simplifying resource management for multi-container pods.

> This work was done as part of [KEP #4678](https://github.com/kubernetes/enhancements/issues/4678) by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).

Reviewer 4: The KEP is https://github.com/kubernetes/enhancements/issues/2837, https://github.com/kubernetes/enhancements/issues/4678 is the PR it seems.

```suggestion
This work was done as part of [KEP #2837](https://github.com/kubernetes/enhancements/issues/2837) by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).
```

- [SIG Docs](https://github.com/kubernetes/community/tree/master/sig-docs) - for the fundamental support in docs and blog reviews and continous collaboration with release Comms and Docs;
- [SIG k8s Infra](https://github.com/kubernetes/community/tree/master/sig-k8s-infra) and [SIG Testing](https://github.com/kubernetes/community/tree/master/sig-testing) - for the outstanding work in keeping the testing framework in check, along with all the infra components necessary;
- [SIG Release](https://github.com/kubernetes/community/tree/master/sig-release) - for the incredible support provided throughout the orchestration of the entire release, addressing challenging issues in a graceful and timely manner;

## Project velocity

The CNCF K8s [DevStats project](https://k8s.devstats.cncf.io/d/11/companies-contributing-in-repository-groups?orgId=1&var-period=m&var-repogroup_name=All) aggregates a number of interesting data points related to the velocity of Kubernetes and various sub-projects. This includes everything from individual contributions to the number of companies that are contributing and is an illustration of the depth and breadth of effort that goes into evolving this ecosystem.

> In the v1.32 release cycle, which ran for 14 weeks (September 9th to December 11th), at the time of writing we verified contributions to Kubernetes from 99 different companies and 506 individuals.
> 
> In the whole Cloud Native ecosystem we have 385 companies counting 2195 total contributors - in line with the [previous release](https://kubernetes.io/blog/2024/08/13/kubernetes-v1-31-release/#project-velocity) cycle! 

Reviewer 5: As discussed separately, I'm making a bit more adjustments to the wording and contents to rightly highlight the data points and moving average.

```suggestion
In the v1.32 release cycle, which ran for 14 weeks (September 9th to December 11th), we saw contributions to Kubernetes from as many as 125 different companies and 559 individuals as of writing.

In the whole Cloud Native ecosystem, the figure goes up to 433 companies counting 2441 total contributors. This sees an increase of 7% more overall contributions compared to the [previous release](https://kubernetes.io/blog/2024/08/13/kubernetes-v1-31-release/#project-velocity) cycle, along with 14% increase in the number of companies involved, showcasing strong interest and community behind the Cloud Native projects.
```

Source for this data: 
- [Companies contributing to Kubernetes](https://k8s.devstats.cncf.io/d/11/companies-contributing-in-repository-groups?orgId=1&from=1725832800000&to=1733961599000&var-period=d28&var-repogroup_name=Kubernetes&var-repo_name=kubernetes%2Fkubernetes)
- [Overall ecosystem contributions](https://k8s.devstats.cncf.io/d/11/companies-contributing-in-repository-groups?orgId=1&from=1725832800000&to=1733961599000&var-period=d28&var-repogroup_name=All&var-repo_name=kubernetes%2Fkubernetes)

By contribution we mean when someone makes a commit, code review, comment, creates an issue or PR, reviews a PR (including blogs and documentation) or comments on issues and PRs.

> If you are interested in contributing visit [this page](https://www.kubernetes.dev/docs/guide/#getting-started) to get started. 

Reviewer: (nit)
I'd write
```suggestion
If you are interested in contributing visit [Getting Started](https://www.kubernetes.dev/docs/guide/#getting-started) on our contributor website.
```

(we avoid “here“ in hyperlinks)

- [**KCD - Kubernetes Community Days: Bangalore, India**](https://www.cncf.io/kcds/): June 6, 2025 | Bangalore, India  
- [**KubeCon + CloudNativeCon China 2025**](https://events.linuxfoundation.org/kubecon-cloudnativecon-china/): June 10-11, 2025 | Hong Kong  
- [**KCD - Kubernetes Community Days: Antigua Guatemala, Guatemala**](https://www.cncf.io/kcds/): June 14, 2025 | Antigua Guatemala, Guatemala  
- [**KubeCon + CloudNativeCon Japan 2025**](https://events.linuxfoundation.org/kubecon-cloudnativecon-japan): June 16-17, 2025 | Tokyo, Japan 
- [**KCD - Kubernetes Community Days: Nigeria, Africa**](https://www.cncf.io/kcds/): June 19, 2025 | Nigeria, Africa  

## Upcoming release webinar (WIP)

Join members of the Kubernetes v1.32 release team on {Day, Mon, dd, YYYY - time AM/PM + timezone}, to learn about the release highlights of this release, as well as deprecations and removals to help plan for upgrades. 
For more information and registration, visit the [event page](https://community.cncf.io/events/details/cncf-cncf-online-programs-presents-cncf-live-webinar-kubernetes-132-release/) on the CNCF Online Programs site.

## Get involved

> The simplest way to get involved with Kubernetes is by joining one of the many [Special Interest Groups](https://github.com/kubernetes/community/blob/master/sig-list.md) (SIGs) that align with your interests. 

Reviewer: Could link to https://www.kubernetes.dev/community/community-groups/#special-interest-groups here

Author: Good idea, done!

---
layout: blog
title: 'Kubernetes v1.32: {release-name}'
date: 2024-12-11
slug: kubernetes-v1-32-release
author: >
  [Kubernetes v1.32 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.32/release-team.md)
draft: true
---

**Editors:** Matteo Bianchi, Edith Puclla, William Rizzo, Ryota Sawada, Rashan Smith

Announcing the release of Kubernetes v1.32: {release-name}!

In line with previous releases, the release of Kubernetes v1.32 introduces new stable, beta, and alpha features. 
The consistent delivery of high-quality releases underscores the strength of our development cycle and the vibrant support from our community.
This release consists of 44 enhancements in total.
Of those enhancements, 13 have graduated to Stable, 12 are entering Beta, and 19 have entered in Alpha.

## Release theme and logo
{{< figure src="/images/blog/2024-12-11-kubernetes-1.32-release/k8s-1.32.png" alt="Kubernetes v1.32 logo" class="release-logo" >}}

<TODO upload image to static/images/blog/2024-12-11-kubernetes-1.32-release/k8s-1.32.png>

The Kubernetes v1.32 Release Theme is "{release-name}".

Kubernetes v1.32's {release-story} 

> 

Reviewer 2: Consider an H2 here so that the DRA and QoL changes aren't children of the release theme and story. 

```suggestion
## Updates to recent key features
```

Author: Agreed, done with the next commit!

### A note on DRA enhancements

> In this release, like the previous one, the Kubernetes project continues proposing a number of enhancements to the Dynamic Resource Allocation (DRA), a key component of the Kubernetes resource management system. These enhancements aim to improve the flexibility and efficiency of resource allocation for workloads that require specialized hardware, such as GPUs, FPGAs and network adapters.

Reviewer 2: ```suggestion
In this release, like in v1.31, the Kubernetes project continues proposing a number of enhancements to the Dynamic Resource Allocation (DRA), a key component of the Kubernetes resource management system. These enhancements aim to improve the flexibility and efficiency of resource allocation for workloads that require specialized hardware, such as GPUs, FPGAs, and network adapters.
```

Author: I prefer "the previous one" formula, since we are in 1.32 the previous must be 1.31 anyway!

> The core part of Dynamic Resource Allocation (DRA), the [structured parameter support](https://github.com/kubernetes/enhancements/issues/4381), got promoted to beta.

Reviewer 2: We already defined the acronym 

```suggestion
The core part of DRA, the [structured parameter support](https://github.com/kubernetes/enhancements/issues/4381), was promoted to beta.
```

Author: Not relevant anymore, paragraph rewritten!

> The core part of Dynamic Resource Allocation (DRA), the [structured parameter support](https://github.com/kubernetes/enhancements/issues/4381), got promoted to beta.

Reviewer 2: Consider bulleted list for these to make it more scannable.

Author: Not relevant anymore, paragraph rewritten!

Publishing information about a device in the ResourceClaim status, in particular for network interfaces, got added as part of KEP [#4817](https://github.com/kubernetes/enhancements/issues/4817) as a new alpha feature by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).

> Publishing resource health status in the Pod status has been designed, as outlined in KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680), but not full implemented yet. This feature will provide a more detailed view of the health status of resources in the Pod, allowing users to better understand the state of their resources and take appropriate action when necessary.

Reviewer 2: nit

```suggestion
Publishing resource health status in the Pod status has been designed, as outlined in KEP [#4680](https://github.com/kubernetes/enhancements/issues/4680), but not fully implemented yet. This feature will provide a more detailed view of the health status of resources in the Pod, allowing users to better understand the state of their resources and take appropriate action when necessary.
```

Author: Not relevant anymore, paragraph rewritten!

> Another new KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816) will add support for prioritized selection criteria associated to a device request in a ResourceClaim.

Reviewer 2: ```suggestion
Support for prioritized selection criteria associated to a device request in a ResourceClaim will be added with KEP [#4816](https://github.com/kubernetes/enhancements/issues/4816)
```

Reviewer 2: parallelism

Author: Not relevant anymore, paragraph rewritten!

> DRA _structured parameters_ will gain the ability to make requests for very specific types of devices using a ResourceClaim. This feature will allow users to request specific devices, such as GPUs or FPGAs, and have the scheduler prioritize the selection of nodes with the requested devices. This feature is particularly useful for workloads that require specialized hardware, such as machine learning or high-performance computing applications.

Reviewer 2: This seems like a repeat of the first point? Merge it into a point that makes sense. Looks a bit weird on its own at the end here when it feels like we touched on it already

Author: Agreed, removed in favor of a small note/comment on it without mentioning the single KEPs

### Quality of life improvements on nodes and sidecar containers update

> Some highlights beyond KEPs to take a look at, by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).

Reviewer 2: ```suggestion
[SIG Node](https://github.com/kubernetes/community/tree/master/sig-node) has the following highlights
that go beyond KEPs:
```

Author: This is mostly a stilistic choice (no real improvement comes from this change) but I'm ok giving this in!

> 1. The systemd watchdog capability is now utilized to restart the kubelet when its health check fails, while limiting the maximum number of restarts within a given time period. This enhances the reliability of the kubelet to some extent. You can find more details in the pull request [#127566](https://github.com/kubernetes/kubernetes/pull/127566).

Reviewer 2: ```suggestion
1. The systemd watchdog capability is now used to restart the kubelet when its health check fails, while
     also limiting the maximum number of restarts within a given time period. This enhances the reliability
     of the kubelet. For more details, see pull request [#127566](https://github.com/kubernetes/kubernetes/pull/127566).
```

Author: Fixed in the upcoming commit

> 2. In cases when an image pull back-off error is encountered, the message displayed in the pod status has been improved to be more human-friendly and to indicate details about why the pod is in this condition.

Reviewer 2: ```suggestion
2. The message that's displayed in the Pod status when image pull back-off errors occur
     is now more human-friendly and includes details about why the Pod is in this condition.
```

Author: Fixed btw!

> The image pull error is now appended to the pods `status.containerStatuses[*].state.waiting.message` when in image pull back-off (`reason` is `ImagePullBackOff`). This change provides more context to the user, making it easier to identify the root cause of the issue. You can find more details in the pull request [#127918](https://github.com/kubernetes/kubernetes/pull/127918).

Reviewer 2: ```suggestion
When an image pull back-off occurs, the error is appended to the `status.containerStatuses[*].state.waiting.message` field in the Pod specification with an
`ImagePullBackOff` value in the `reason` field. This change provides you with more context
and helps you to identify the root cause of the issue. For more details, see pull request [#127918](https://github.com/kubernetes/kubernetes/pull/127918).
```

Author: Fixed in the upcoming commit

> 3. The Sidecar containers feature is targeting graduation to Stable in 1.33 as the [remaining work items](https://github.com/kubernetes/enhancements/issues/753#issuecomment-2350136594) are being tracked along with feedback from users.

Reviewer 2: ```suggestion
3. The sidecar containers feature is targeting graduation to Stable in v1.33. To view the remaining work
     items and feedback from users, see the comments in issue [#753](https://github.com/kubernetes/enhancements/issues/753#issuecomment-2350136594).
```

Author: Fixed with the upcoming commit

> ## Highlights of features graduating to Stable

Reviewer 2: ```suggestion
## Features graduating to Stable
```

Reviewer 2: Mirror other headings for beta and alpha

Author: Thanks for pointing it out, it's all a sub-selection so I actually fixed all the other headings

_This is a selection of some of the improvements that are now stable following the v1.32 release._

### Custom Resource field selectors

Custom resource field selector allows developers to add field selectors to custom resources, mirroring the functionality available for built-in Kubernetes objects. This allows for more efficient and precise filtering of custom resources, promoting better API design practices.

This work was done as a part of [KEP #4358](https://github.com/kubernetes/enhancements/issues/4358), by [SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery).

> ### Support to size memory backed volumes

Reviewer 2: ```suggestion
### Dynamic sizing of memory backed volumes
```

Author: This is how the KEP is titled, unsure if we can change it - I'll refuse the change for now

> The feature makes it possible to dynamically size memory-backed volumes based on pod resource limits. This improves the workload's portability and overall node resource utilization for the users.

Reviewer 2: ```suggestion
This feature makes it possible to dynamically size memory-backed volumes based on Pod resource limits. This improves the workload's portability and overall node resource utilization.
```

Author: Fixed with the next commit

    
This work was done as a part of [KEP #1967](https://github.com/kubernetes/enhancements/issues/1967), by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).

### Job API managed-by mechanism

> Support of the `managedBy` field has been added for Jobs and it is stable in 1.32 release. This feature enables external controllers (like [Kueue](https://kueue.sigs.k8s.io/)) to manage Job synchronization, offering greater flexibility and integration with advanced workload management systems.

Reviewer 2: ```suggestion
The `managedBy` field has been added for Jobs and is stable in v1.32. This feature enables external controllers like [Kueue](https://kueue.sigs.k8s.io/) to manage Job synchronization, offering greater flexibility and integration with advanced workload management systems.
```

Author: Partially rewritten in the upcoming commit!

> Support of the `managedBy` field has been added for Jobs and it is stable in 1.32 release. This feature enables external controllers (like [Kueue](https://kueue.sigs.k8s.io/)) to manage Job synchronization, offering greater flexibility and integration with advanced workload management systems.

Reviewer 6: ```suggestion
Support of the `managedBy` field has been added for Jobs and it was promoted to beta and on by default in 1.32 release. This feature enables external controllers (like [Kueue](https://kueue.sigs.k8s.io/)) to manage Job synchronization, offering greater flexibility and integration with advanced workload management systems.
```

Reviewer 6: We should probably also move this below to the beta list.

Author: Thank you, it will be fixed on the upcoming commit!

This work was done as a part of [KEP #4368](https://github.com/kubernetes/enhancements/issues/4368), by [SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps).

## Features graduating to Beta

_This is a selection of some of the improvements that are now beta following the v1.32 release._

### Only allow anonymous auth for configured endpoints

> Allow users to specify which endpoints are allowed for anonymous requests. This allows the admin to only allow access to health endpoints like `/healthz`, `/livez` and `/readyz` anonymously while making sure other cluster endpoints or resources cannot be access anonymously even if a user misconfigures RBAC.

Reviewer 2: ```suggestion
This feature lets admins specify which endpoints are allowed for anonymous requests. For example, the admin
can choose to only allow anonymous access to health endpoints like `/healthz`, `/livez`, and `/readyz` while
making sure preventing anonymous access to other cluster endpoints or resources even if a user
misconfigures RBAC.
```

Author: Fixed in the upcoming commit!

    
This work was done as a part of [KEP #4633](https://github.com/kubernetes/enhancements/issues/4633), by [SIG Auth](https://github.com/kubernetes/community/tree/master/sig-auth).

### Per-plugin callback functions for accurate requeueing in kube-scheduler enhancements

> Enhance the scheduling throughput with more efficient scheduling retry decisions by per-plugin callback functions (QueueingHint). 
> We've implemented QueueingHints in all plugins, and are moving ahead enabling this feature by default.

Reviewer 2: ```suggestion
This feature enhance sthe scheduling throughput with more efficient scheduling retry decisions by
per-plugin callback functions (QueueingHint). All plugins now have QueueingHints.
```
Cut the bit that's true for all beta stuff

Author: Done!

This work was done as a part of [KEP #4247](https://github.com/kubernetes/enhancements/issues/4247), by [SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling).

### Recover from volume expansion failure

> This feature moved to Beta in v1.32, allowing users to recover from volume expansion failure by retrying with a smaller size. This enhancement ensures that volume expansion is more resilient and reliable, reducing the risk of data loss or corruption during the process.

Reviewer 2: ```suggestion
This feature lets users recover from volume expansion failure by retrying with a smaller size. This enhancement ensures that volume expansion is more resilient and reliable, reducing the risk of data loss or corruption during the process.
```

Author: Fixed!

This work was done as a part of [KEP #1790](https://github.com/kubernetes/enhancements/issues/1790), by [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

### Volume group snapshot

> This feature introduces a VolumeGroupSnapshot API to take a snapshot of multiple volumes together.
> Allows users to take a snapshot of multiple volumes together, ensuring data consistency across the volumes.

Reviewer 2: ```suggestion
This feature introduces a VolumeGroupSnapshot API to take a snapshot of multiple volumes together to
ensure data consistency across the volumes.
```

First half of second sentence is a repeat of first sentence

Author: Whoops, fixed.

This work was done as a part of [KEP #3476](https://github.com/kubernetes/enhancements/issues/3476), by [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

### Structured parameter support

The core part of Dynamic Resource Allocation (DRA), the structured parameter support, got promoted to beta.

> Thanks to this feature, the kube-scheduler and Cluster Autoscaler can handle respectively simulate claim allocation themselves without relying on a third-party driver.

Reviewer 2: "can handle respectively simulate claim allocation themselves" 

I can't understand this at all, not even after reading the actual KEP for structured parameters :/ Can we make this clearer?

Author: I can try but keep in mind that this is the text we got from the KEP itself and not being part of that specific SIG I have not much extra information to rely on when trying and simplifying it.

Did my best in the latest version - see upcoming commit -

This work was done as a part of [KEP #4381](https://github.com/kubernetes/enhancements/issues/4381), by WG Device Management (a cross functional team containing [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node), [SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling) and [SIG Autoscaling](https://github.com/kubernetes/community/tree/master/sig-autoscaling)).

## New features in Alpha

_This is a selection of key improvements introduced as alpha features in the v1.32 release._
    
### Asynchronous preemption in the Kubernetes Scheduler

> Enhance the scheduling throughput when the preemption is involved, by making some expensive operations within the preemption asynchronously.

Reviewer 2: ```suggestion
This feature enhances the scheduling throughput when preemption is involved by making some expensive
preemption operations asynchronous.
```

Be good if we could say what we mean by "expensive"

Reviewer 7: Thanks @shannonxtreme for this feedback.
"Expensive operations" are not explicitly in the enhancement description and is a term to cover: time-consuming, scheduler performance

We can do two things :
1. Clarify it with an example we can dd in parenthesis.
`expensive operations (e.g., resource-intensive API calls like updating pod conditions and deleting pods, which can delay scheduling cycles)`

2. We can also change that phrase and add `Resource-intensive operations` or `Heavy operations` instead of "Expensive operations" to refer to all high resource demands, such as CPU or network calls, mentioned in the enhancement.

What do you think? 🙂

Reviewer 2: I like resource-intensive :)

Author: Great, I'll use that instead!

- want to fix every of your comments in a single commit so I'm working local for now @shannonxtreme 
Thanks for all the great feedback, much appreciated

This improvement is particularly beneficial in clusters with high pod churn or frequent scheduling failures, ensuring a more efficient and resilient scheduling process.

> The Kubernetes scheduler has been enhanced with Asynchronous Preemption, a feature that improves scheduling throughput by handling preemption operations asynchronously. Preemption ensures higher-priority pods get the resources they need by evicting lower-priority ones, but this process previously involved blocking operations like PostFilter execution, slowing down the scheduler. With this enhancement, such tasks are now processed in parallel, allowing the scheduler to continue scheduling other pods without delays.

Reviewer 2: This is awesome, I would cut the previous paragraph and go with this one alone

Author: Fixed, just kept the reason why the improvement is beneficial at the end

This work was done as a part of KEP [#4832](https://github.com/kubernetes/enhancements/issues/4832) by [SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling).

 
### Mutating admission policies using CEL expressions

This feature leverages CEL's object instantiation and JSON Patch strategies, combined with Server Side Apply’s merge algorithms. It simplifies policy definition, reduces mutation conflicts, and enhances admission control performance while laying a foundation for more robust, extensible policy frameworks in Kubernetes.

> The Kubernetes scheduler now supports CEL-based Mutating Admission Policies, providing a lightweight, efficient alternative to mutating admission webhooks. With this enhancement, administrators can use the Common Expression Language (CEL) to declare mutations like setting labels, defaulting fields, or injecting sidecars with simple, declarative expressions. This approach reduces operational complexity, eliminates the need for webhooks, and integrates directly with the kube-apiserver, offering faster and more reliable in-process mutation handling.

Reviewer 2: ```suggestion
The Kubernetes scheduler now supports Common Expression Language (CEL)-based Mutating Admission Policies, providing a lightweight, efficient alternative to mutating admission webhooks. With this enhancement, administrators can use CEL to declare mutations like setting labels, defaulting fields, or injecting sidecars with simple, declarative expressions. This approach reduces operational complexity, eliminates the need for webhooks, and integrates directly with the kube-apiserver, offering faster and more reliable in-process mutation handling.
```

Author: Fixed!

This work was done as a part of [KEP #3962](https://github.com/kubernetes/enhancements/issues/3962) by [SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery).

> ### Pod-level Resource specifications 

Reviewer 2: ```suggestion
### Pod-level resource specifications 
```

Author: Fixed!

> This enhancement introduces a simplified way to manage resources in Kubernetes by allowing users to define resource requests and limits for an entire pod instead of specifying them individually for each container. This creates a shared resource pool that all containers in the pod can dynamically utilize, optimizing resource usage and reducing over-provisioning, especially for workloads with varying or bursty resource demands.

Reviewer 2: ```suggestion
This enhancement introduces a centralized way to manage resources in Kubernetes by allowing users to define resource requests and limits for an entire Pod instead of specifying them individually for each container. This creates a shared resource pool that all containers in the Pod can dynamically utilize, optimizing resource usage and reducing over-provisioning, especially for workloads with varying or bursty resource demands.
```

Author: Completely rewritten, PTAL (see below)

> With this feature, users gain more flexibility and efficiency in resource allocation, as the pod-level settings complement existing container-level specifications. By directly leveraging pod-level cgroup settings, Kubernetes ensures better performance for tightly coupled applications while maintaining backward compatibility with container-level configurations. This enhancement represents a significant step forward in simplifying resource management for multi-container pods.

Reviewer 2: This paragraph feels like it says cool things without saying why or how they're cool. Why would I need backward compat with container-level resource settings? What does cgroup have to do with this? Why is this such a huge step forward for multi-container pods? 

I had all of the above Qs after reading this paragraph, but I don't feel like I got any answers in it

Author: Rewritten in the upcoming commit, PTAL

The decision to change the existing approach originated from its incompatibility with cluster autoscaling as resource availability was non-transparent, complicating decision-making for both Cluster Autoscaler and controllers. 
The newly added Structured Parameter model substitutes the functionality.

This removal will allow Kubernetes to handle new hardware requirements and resource claims more predictably, bypassing the complexities of back and forth API calls to the kube-apiserver.

Please also see the enhancement issue [#3063](https://github.com/kubernetes/enhancements/issues/3063) to find out more.

#### Deprecation of gitRepo volume types

The [gitRepo](https://kubernetes.io/docs/concepts/storage/volumes/#gitrepo) volume type is deprecated and will be removed in a future release, the deprecation has been executed in light of the security advisory encompassing the [CVE-2024-10220](https://nvd.nist.gov/vuln/detail/CVE-2024-10220): Arbitrary command execution through gitRepo volume, which was reported publicly in [this issue](https://github.com/kubernetes/kubernetes/issues/128885).
    
#### API removals

> There is only a single API removal planned for [Kubernetes v1.32](/docs/reference/using-api/deprecation-guide/#v1-32):

Reviewer 2: ```suggestion
There is one API removal in [Kubernetes v1.32](/docs/reference/using-api/deprecation-guide/#v1-32):
```

Reviewer 2: Blog is for 1.32, removal will no longer be planned

Author: Fixed in the next commit!

* The `flowcontrol.apiserver.k8s.io/v1beta3` API version of FlowSchema and PriorityLevelConfiguration has been removed. 
To prepare for this, you can edit your existing manifests and rewrite client software to use the `flowcontrol.apiserver.k8s.io/v1 API` version, available since v1.29. 
All existing persisted objects are accessible via the new API. Notable changes in flowcontrol.apiserver.k8s.io/v1beta3 include that the PriorityLevelConfiguration `spec.limited.nominalConcurrencyShares` field only defaults to 30 when unspecified, and an explicit value of 0 is not changed to 30.

> For more information, please refer to the [API deprecation guide](/docs/reference/using-api/deprecation-guide/#v1-32).

Reviewer 2: here and elsewhere, cut "please"

Author: Cut the 2 (two) instances of "please" in the next commit!

support from our community.
This release consists of 44 enhancements in total.
Of those enhancements, 13 have graduated to Stable, 12 are entering Beta, and 19 have entered in Alpha.

## Release theme and logo
{{< figure src="/images/blog/2024-12-11-kubernetes-1.32-release/k8s-1.32.png" alt="Kubernetes v1.32 logo"
class="release-logo" >}}

<TODO upload image to static/images/blog/2024-12-11-kubernetes-1.32-release/k8s-1.32.png>

The Kubernetes v1.32 Release Theme is "{release-name}".

Kubernetes v1.32's {release-story}

## Updates to recent key features

### A note on DRA enhancements

In this release, like the previous one, the Kubernetes project continues proposing a number of enhancements to the
Dynamic Resource Allocation (DRA), a key component of the Kubernetes resource management system. These enhancements aim
to improve the flexibility and efficiency of resource allocation for workloads that require specialized hardware, such
as GPUs, FPGAs and network adapters.
These features are particularly useful for use-cases such as machine learning or high-performance computing

> applications.

Reviewer 8: Since the section “**Structured parameter support** ”(Line 129) is also an important update to DRA in this release, would it be better to move it here? It is more convenient for users to get key information quickly.

Or don't move the whole paragraph and mention here that the “**Highlights of features graduating to Beta**” section has more details about “**Structured parameter support** ”? IMO.

Author: I didn't move it on purpose, before this version we had the entire list of DRA enhancements in there and it was a bit too much.

I added an anchor link that should help the user rapidly navigate to the relative highlight!

Thank you for the input :)

Reviewer 8: > I added an anchor link that should help the user rapidly navigate to the relative highlight!

I agree with this approach.

by handling preemption operations asynchronously. Preemption ensures higher-priority pods get the resources they need by
evicting lower-priority ones, but this process previously involved blocking operations like PostFilter execution,
slowing down the scheduler. With this enhancement, such tasks are now processed in parallel, allowing the scheduler to
continue scheduling other pods without delays.
This improvement is particularly beneficial in clusters with high Pod churn or frequent scheduling failures, ensuring a
more efficient and resilient scheduling process.

This work was done as a part of KEP [#4832](https://github.com/kubernetes/enhancements/issues/4832) by [SIG
Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling).

 
### Mutating admission policies using CEL expressions

This feature leverages CEL's object instantiation and JSON Patch strategies, combined with Server Side Apply’s merge
algorithms. It simplifies policy definition, reduces mutation conflicts, and enhances admission control performance
while laying a foundation for more robust, extensible policy frameworks in Kubernetes.

> The Kubernetes scheduler now supports Common Expression Language (CEL)-based Mutating Admission Policies, providing a

Reviewer 9: maybe:  scheduler -> apiserver

Reviewer: Good point, but: “API server“ not “apiserver“

Reviewer 8: > maybe: scheduler -> apiserver

+1, I was also confused when I saw this, both [KEP](https://github.com/kubernetes/enhancements/issues/3962) and [code implementation](https://github.com/kubernetes/kubernetes/pull/127134) are related to the `API server `component, not the scheduler component.

Author: Fixed, thanks for catching this!

applications.

### Quality of life improvements on nodes and sidecar containers update

[SIG Node](https://github.com/kubernetes/community/tree/master/sig-node) has the following highlights that go beyond
KEPs:

1. The systemd watchdog capability is now used to restart the kubelet when its health check fails, while also limiting
the maximum number of restarts within a given time period. This enhances the reliability of the kubelet. For more
details, see pull request [#127566](https://github.com/kubernetes/kubernetes/pull/127566).

2. In cases when an image pull back-off error is encountered, the message displayed in the Pod status has been improved
to be more human-friendly and to indicate details about why the Pod is in this condition.
When an image pull back-off occurs, the error is appended to the `status.containerStatuses[*].state.waiting.message`
field in the Pod specification with an `ImagePullBackOff` value in the `reason` field. This change provides you with
more context and helps you to identify the root cause of the issue. For more details, see pull request
[#127918](https://github.com/kubernetes/kubernetes/pull/127918).

> 1. The sidecar containers feature is targeting graduation to Stable in v1.33. To view the remaining work items and

Reviewer 2: The numbering goes 1. 2. 1., maybe switch all to `1.`? Otherwise might render oddly, not sure

Author: Just a 3. typo, fixed with the next commit - that will be squashed/rebased if needed

_This is a selection of some of the improvements that are now stable following the v1.32 release._

### Custom Resource field selectors

Custom resource field selector allows developers to add field selectors to custom resources, mirroring the functionality
available for built-in Kubernetes objects. This allows for more efficient and precise filtering of custom resources,
promoting better API design practices.

This work was done as a part of [KEP #4358](https://github.com/kubernetes/enhancements/issues/4358), by [SIG API
Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery).

### Support to size memory backed volumes

This feature makes it possible to dynamically size memory-backed volumes based on Pod resource limits, improving the
workload's portability and overall node resource utilization.

This work was done as a part of [KEP #1967](https://github.com/kubernetes/enhancements/issues/1967), by [SIG
Node](https://github.com/kubernetes/community/tree/master/sig-node).

## Highlights of features graduating to Beta

_This is a selection of some of the improvements that are now beta following the v1.32 release._

### Job API managed-by mechanism

> The `managedBy` field for Jobs and was promoted to beta in the v1.32 release. This feature enables external controllers

Reviewer 2: ```suggestion
The `managedBy` field for Jobs was promoted to beta in the v1.32 release. This feature enables external controllers
```

(like [Kueue](https://kueue.sigs.k8s.io/)) to manage Job synchronization, offering greater flexibility and integration
with advanced workload management systems.

This work was done as a part of [KEP #4368](https://github.com/kubernetes/enhancements/issues/4368), by [SIG
Apps](https://github.com/kubernetes/community/tree/master/sig-apps).

### Only allow anonymous auth for configured endpoints

This feature lets admins specify which endpoints are allowed for anonymous requests. For example, the admin
can choose to only allow anonymous access to health endpoints like `/healthz`, `/livez`, and `/readyz` while
making sure preventing anonymous access to other cluster endpoints or resources even if a user
misconfigures RBAC.
    
This work was done as a part of [KEP #4633](https://github.com/kubernetes/enhancements/issues/4633), by [SIG
Auth](https://github.com/kubernetes/community/tree/master/sig-auth).

### Per-plugin callback functions for accurate requeueing in kube-scheduler enhancements

> This feature enhance sthe scheduling throughput with more efficient scheduling retry decisions by

Reviewer 2: ```suggestion
This feature enhances scheduling throughput with more efficient scheduling retry decisions by
```

Author: Whoops, fixed!

per-plugin callback functions (QueueingHint). All plugins now have QueueingHints.

This work was done as a part of [KEP #4247](https://github.com/kubernetes/enhancements/issues/4247), by [SIG
Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling).

### Recover from volume expansion failure

This feature lets users recover from volume expansion failure by retrying with a smaller size. This enhancement ensures
that volume expansion is more resilient and reliable, reducing the risk of data loss or corruption during the process.

This work was done as a part of [KEP #1790](https://github.com/kubernetes/enhancements/issues/1790), by [SIG
Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

### Volume group snapshot

> This feature introduces a VolumeGroupSnapshot API.
> Allows users to take a snapshot of multiple volumes together, ensuring data consistency across the volumes.

Reviewer 2: ```suggestion
This feature introduces a VolumeGroupSnapshot API, which lets users take a snapshot of multiple volumes together, ensuring data consistency across the volumes.
```

Author: Fixed!

This work was done as a part of [KEP #3476](https://github.com/kubernetes/enhancements/issues/3476), by [SIG
Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

### Structured parameter support

The core part of Dynamic Resource Allocation (DRA), the structured parameter support, got promoted to beta.
This allows the kube-scheduler and Cluster Autoscaler to simulate claim allocation directly, without needing a
third-party driver.

> Maning they can predict whether resource requests can be fulfilled based on the cluster's current state without actually

Reviewer 2: ```suggestion
These components can now predict whether resource requests can be fulfilled based on the cluster's current state without actually
```

Author: Fixed and accepted suggestion (it was meant to be "meaning" ofc)

as sidecar architectures, where containers share workloads or depend on each other’s availability to perform optimally.

This work was done as part of [KEP #2837](https://github.com/kubernetes/enhancements/issues/2837) by [SIG
Node](https://github.com/kubernetes/community/tree/master/sig-node).

    
### Allow zero value for sleep action of PreStop hook

This enhancement introduces the ability to set a zero-second sleep duration for the PreStop lifecycle hook in
Kubernetes, offering a more flexible and no-op option for resource validation and customization. Previously, attempting
to define a zero value for the sleep action resulted in validation errors, restricting its use. With this update, users
can configure a zero-second duration as a valid sleep setting, enabling immediate execution and termination behaviors
where needed.

The enhancement is backward-compatible, introduced as an opt-in feature controlled by the
`PodLifecycleSleepActionAllowZero` feature gate. This change is particularly beneficial for scenarios requiring PreStop
hooks for validation or admission webhook processing without requiring an actual sleep duration. By aligning with the

> capabilities of Go’s time.After function, this update simplifies configuration and expands usability for Kubernetes

Author: Fixed!
